File: fsharp/protein-translation/ProteinTranslation.fs
module ProteinTranslation


let of_codon codon =
    match codon with
    | "UGU" -> "Cysteine"
    | "UGC" -> "Cysteine"
    | "UUA" -> "Leucine"
    | "UUG" -> "Leucine"
    | "AUG" -> "Methionine"
    | "UUU" -> "Phenylalanine"
    | "UUC" -> "Phenylalanine"
    | "UCU" -> "Serine"
    | "UCC" -> "Serine"
    | "UCA" -> "Serine"
    | "UCG" -> "Serine"
    | "UGG" -> "Tryptophan"
    | "UAU" -> "Tyrosine"
    | "UAC" -> "Tyrosine"
    | "UAA" -> "STOP"
    | "UAG" -> "STOP"
    | "UGA" -> "STOP"
    | _ -> ""

let rec do_translate (input, acc) = 

    match input with
    | [] -> acc
    | x :: xs -> 
        match of_codon x with 
        | "STOP" -> acc
        | codon -> do_translate(xs, acc @ [codon])

let translate(input) = do_translate(input, [])

let proteins rna = 

    match rna with
    | "" -> []
    | protein -> protein 
                            |> Seq.toList 
                            |> Seq.chunkBySize(3) 
                            |> Seq.map(fun s -> new string(s)) 
                            |> Seq.toList
                            |> translate
    


 // let proteins rna = failwith "You need to implement this function."

File: c/protein-translation/protein_translation.c
#include "protein_translation.h"

#include <string.h>

protein_t translate(const char *codon);

proteins_t proteins(const char *const rna) {
    
    if (strcmp(rna, "") == 0) {
        return (proteins_t) {
            .valid = true,
            .count = 0,
        };
    }

    size_t count = 0;
    proteins_t res = (proteins_t) {
        .valid = true,
        .count = 0,
    };

    char codon[4];
    int len = strlen(rna);

    for (int off = 0; off < len; off += 3) {
        strncpy(codon, rna + off, 3);
        codon[3] = '\0';

        protein_t protein = translate(codon);

        if (protein == INVALID) {
            return (proteins_t) {
                .valid = false,
            };
        }

        if (protein == STOP) {
            return res;
        }

        res.proteins[count++] = protein;
        res.count = count;
    }

    return res;
}

protein_t translate(const char *codon) {
    
    if (strcmp(codon, "AUG") == 0) {
        return Methionine;
    }

    if (strcmp(codon, "UUU") == 0 || strcmp(codon, "UUC") == 0) {
        return Phenylalanine;
    }

    if (strcmp(codon, "UUA") == 0 || strcmp(codon, "UUG") == 0) {
        return Leucine;
    }

    if (strcmp(codon, "UCU") == 0 || strcmp(codon, "UCC") == 0
         || strcmp(codon, "UCA") == 0 || strcmp(codon, "UCG") == 0) {
        return Serine;
    }

    if (strcmp(codon, "UAU") == 0 || strcmp(codon, "UAC") == 0) {
        return Tyrosine;
    }

    if (strcmp(codon, "UGU") == 0 || strcmp(codon, "UGC") == 0) {
        return Cysteine;
    }

    if (strcmp(codon, "UGG") == 0) {
        return Tryptophan;
    }
    
    if (strcmp(codon, "UAA") == 0 || strcmp(codon, "UAG") == 0
         || strcmp(codon, "UGA") == 0) {
        return STOP;
    }

    return INVALID;

}
